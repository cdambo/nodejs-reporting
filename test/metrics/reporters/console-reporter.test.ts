import { invoke } from "lodash";
import ConsoleReporter from "../../../src/metrics/reporters/console-reporter";

const getReporter = (): ConsoleReporter =>
  new ConsoleReporter({
    globalTags: { globalTag1: "global_one", globalTag2: "global_two" }
  });

const reportsOf = (
  method: keyof ConsoleReporter,
  ...args: unknown[]
): ReturnType<typeof jest.spyOn> => {
  const reporter = getReporter();
  const mockOutput = jest.spyOn(reporter, "output").mockImplementation();

  invoke(reporter, method, ...args);
  return mockOutput.mock.calls;
};

describe("ConsoleReporter", (): void => {
  describe("Constructor", (): void => {
    it("Creates a ConsoleReporter", (): void => {
      const reporter = new ConsoleReporter();
      expect(reporter).toMatchSnapshot();
    });

    it("Creates a ConsoleReporter with global tags", (): void => {
      const reporter = new ConsoleReporter({
        globalTags: { globalTag1: "global_one", globalTag2: "global_two" }
      });
      expect(reporter).toMatchSnapshot();
    });

    it("Creates a ConsoleReporter with a format", (): void => {
      const format = { format: (): string => "Format" };
      const reporter = new ConsoleReporter({ format });
      expect(reporter).toMatchSnapshot();
    });

    it("Creates a ConsoleReporter with an output", (): void => {
      const output = (): void => {};
      const reporter = new ConsoleReporter({ output });
      expect(reporter).toMatchSnapshot();
    });
  });

  describe("timing", (): void => {
    it("Reports timing metric", (): void => {
      expect(
        reportsOf("timing", "timing_test", 20, 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("increment", (): void => {
    it("Reports counter metric", (): void => {
      expect(
        reportsOf("increment", "increment_test", 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("incrementBy", (): void => {
    it("Reports counter metric", (): void => {
      expect(
        reportsOf("incrementBy", "increment_by_test", 10, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("decrement", (): void => {
    it("Reports counter metric", (): void => {
      expect(
        reportsOf("decrement", "decrement_test", 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("decrementBy", (): void => {
    it("Reports counter metric", (): void => {
      expect(
        reportsOf("decrementBy", "decrement_by_test", 10, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("gauge", (): void => {
    it("Reports gauge metric", (): void => {
      expect(
        reportsOf("gauge", "gauge_test", 10, 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("histogram", (): void => {
    it("Reports histogram metric", (): void => {
      expect(
        reportsOf("histogram", "histogram_test", 20, 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });
});
