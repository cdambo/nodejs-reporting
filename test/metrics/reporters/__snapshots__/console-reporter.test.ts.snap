// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsoleReporter Constructor Creates a ConsoleReporter 1`] = `
ConsoleReporter {
  "context": Context {
    "globalTags": Object {},
  },
  "format": [Function],
  "output": [Function],
}
`;

exports[`ConsoleReporter Constructor Creates a ConsoleReporter with a format 1`] = `
ConsoleReporter {
  "context": Context {
    "globalTags": Object {},
  },
  "format": [Function],
  "output": [Function],
}
`;

exports[`ConsoleReporter Constructor Creates a ConsoleReporter with an output 1`] = `
ConsoleReporter {
  "context": Context {
    "globalTags": Object {},
  },
  "format": [Function],
  "output": [Function],
}
`;

exports[`ConsoleReporter Constructor Creates a ConsoleReporter with global tags 1`] = `
ConsoleReporter {
  "context": Context {
    "globalTags": Object {
      "globalTag1": "global_one",
      "globalTag2": "global_two",
    },
  },
  "format": [Function],
  "output": [Function],
}
`;

exports[`ConsoleReporter decrement Reports a counter metric 1`] = `
Array [
  Array [
    "{ DECREMENT: { stat: [32m'decrement_test'[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter decrement Reports a counter metric with a sample rate 1`] = `
Array [
  Array [
    "{ DECREMENT: { stat: [32m'decrement_test'[39m, sampleRate: [33m0.5[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter decrement Reports a counter metric with tags 1`] = `
Array [
  Array [
    "{ DECREMENT: { stat: [32m'decrement_test'[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m, tag_1: [32m'one'[39m, tag_2: [32m'two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter decrementBy Reports a counter metric 1`] = `
Array [
  Array [
    "{ DECREMENT: { stat: [32m'decrement_by_test'[39m, value: [33m10[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter decrementBy Reports a counter metric with tags 1`] = `
Array [
  Array [
    "{ DECREMENT: { stat: [32m'decrement_by_test'[39m, value: [33m10[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m, tag_1: [32m'one'[39m, tag_2: [32m'two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter gauge Reports a gauge metric 1`] = `
Array [
  Array [
    "{ GAUGE: { stat: [32m'gauge_test'[39m, value: [33m10[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter gauge Reports a gauge metric with a sample rate 1`] = `
Array [
  Array [
    "{ GAUGE: { stat: [32m'gauge_test'[39m, value: [33m10[39m, sampleRate: [33m0.5[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter gauge Reports a gauge metric with tags 1`] = `
Array [
  Array [
    "{ GAUGE: { stat: [32m'gauge_test'[39m, value: [33m10[39m, sampleRate: [33m0.5[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m, tag_1: [32m'one'[39m, tag_2: [32m'two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter histogram Reports a histogram metric 1`] = `
Array [
  Array [
    "{ HISTOGRAM: { stat: [32m'histogram_test'[39m, time: [33m20[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter histogram Reports a histogram metric with a sample rate 1`] = `
Array [
  Array [
    "{ HISTOGRAM: { stat: [32m'histogram_test'[39m, time: [33m20[39m, sampleRate: [33m0.5[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter histogram Reports a histogram metric with tags 1`] = `
Array [
  Array [
    "{ HISTOGRAM: { stat: [32m'histogram_test'[39m, time: [33m20[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m, tag_1: [32m'one'[39m, tag_2: [32m'two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter increment Reports a counter metric 1`] = `
Array [
  Array [
    "{ INCREMENT: { stat: [32m'increment_test'[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter increment Reports a counter metric with a sample rate 1`] = `
Array [
  Array [
    "{ INCREMENT: { stat: [32m'increment_test'[39m, sampleRate: [33m0.5[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter increment Reports a counter metric with tags 1`] = `
Array [
  Array [
    "{ INCREMENT: { stat: [32m'increment_test'[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m, tag_1: [32m'one'[39m, tag_2: [32m'two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter incrementBy Reports a counter metric 1`] = `
Array [
  Array [
    "{ INCREMENT: { stat: [32m'increment_by_test'[39m, value: [33m10[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter incrementBy Reports a counter metric with tags 1`] = `
Array [
  Array [
    "{ INCREMENT: { stat: [32m'increment_by_test'[39m, value: [33m10[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m, tag_1: [32m'one'[39m, tag_2: [32m'two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter timing Reports a timing metric 1`] = `
Array [
  Array [
    "{ TIME: { stat: [32m'timing_test'[39m, time: [33m20[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter timing Reports a timing metric with a sample rate 1`] = `
Array [
  Array [
    "{ TIME: { stat: [32m'timing_test'[39m, time: [33m20[39m, sampleRate: [33m0.5[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m } } }",
  ],
]
`;

exports[`ConsoleReporter timing Reports a timing metric with tags 1`] = `
Array [
  Array [
    "{ TIME: { stat: [32m'timing_test'[39m, time: [33m20[39m, sampleRate: [33m1[39m, tags: { global_tag_1: [32m'global_one'[39m, global_tag_2: [32m'global_two'[39m, tag_1: [32m'one'[39m, tag_2: [32m'two'[39m } } }",
  ],
]
`;
